### PR TITLE
feat(be/search_engine): implement scopus endpoint

### PR DESCRIPTION
**Pull request by jsyska**

### CODE REVIEW

File: rms/external_services/clients.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            "X-ELS-APIKey": self.settings.scopus_api_key
+        }
+        self.SCOPUS_SEARCH_API_ENDPOINT = self.settings.scopus_search_endpoint
+
+    @abstractmethod
+    def search(self, query: SearchBody) -> ScopusSearchResponse:
+        pass
+
+
+class ScopusAPIClient(ScopusClient):
```
Comment (CreativePhilip) [inline rms/external_services/clients.py:22]: I would rename this to `ScopusArticleSearchApi`

File: rms/external_services/tests/test_client.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+app = FastAPI()
+app.include_router(router, prefix="/external_services")
+
+
+@pytest.fixture
+def test_client():
+    return TestClient(app)
+
+
```
Comment (CreativePhilip) [inline rms/external_services/tests/test_client.py:15]: As discussed in person remove this

File: rms/external_services/models/models.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+from pydantic import BaseModel
+from typing import Optional
+
+
+class Affiliation(BaseModel):
+    url: Optional[str]
```
Comment (CreativePhilip) [inline rms/external_services/models/models.py:6]:
```suggestion
    url: str | None
```

File: rms/external_services/tests/test_client.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+def test_search_scopus(test_client):
```
Comment (CreativePhilip) [inline rms/external_services/tests/test_client.py:16]: As discussed please mock the outbound api calls

**Review by CreativePhilip — CHANGES_REQUESTED** [review]

**Review by jankrokos — CHANGES_REQUESTED** [review]

File: rms/external_services/clients.py (hunk 1 above)
Comment (jsyska) [reply rms/external_services/clients.py]: fixed

File: rms/external_services/models/models.py (hunk 3 above)
Comment (jsyska) [reply rms/external_services/models/models.py]: fixed

File: rms/external_services/models/models.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    eid: Optional[str]
+    title: Optional[str]
+    creator: Optional[str]
+    publication_name: Optional[str]
+    cited_by_count: Optional[str]
+    cover_date: Optional[str]
+    self_url: Optional[str]
+    author_affiliation_url: Optional[str]
+    scopus_url: Optional[str]
+    scopus_citedby_url: Optional[str]
```
[the remark this answers is not in the document]
Comment (jsyska) [reply rms/external_services/models/models.py]: fixed

File: rms/external_services/tests/test_client.py (hunk 4 above)
Comment (jsyska) [reply rms/external_services/tests/test_client.py]: fixed

File: rms/external_services/tests/test_client.py (hunk 2 above)
Comment (jsyska) [reply rms/external_services/tests/test_client.py]: fixed

File: rms/search_engine/clients.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        self.HEADERS = {
+            "httpAccept": "application/json",
+            "X-ELS-APIKey": self.settings.scopus_api_key
+        }
+        self.SCOPUS_SEARCH_API_ENDPOINT = self.settings.scopus_search_endpoint
+
+    @abstractmethod
+    def search(self, query: SearchBody) -> ScopusSearchResponse:
+        pass
+
```
Comment (CreativePhilip) [inline rms/search_engine/clients.py:20]: I would remove this as the search params and response types will vary depending on the endpoint

**Review by jankrokos — APPROVED** [review]

File: rms/search_engine/clients.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+class ScopusArticleSearchApi(ScopusClient):
+
+    @staticmethod
+    def build_params(query: SearchBody) -> (str, dict):
```
Comment (CreativePhilip) [inline rms/search_engine/clients.py:25]:
```suggestion
    def build_params(query: SearchBody) -> dict[str, Any]:
```

File: rms/search_engine/clients.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            "view": "COMPLETE",
+            "sort": "-relevancy,-citedby_count,-coverDate"
+        }
+
+        return params
+
+    def search(self, query: SearchBody) -> ScopusSearchResponse:
+        params = self.build_params(query)
+
+        with httpx.Client() as client:
```
Comment (CreativePhilip) [inline rms/search_engine/clients.py:50]: Please change this to use the async client

File: rms/search_engine/models/scopus_models.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class ScopusSearch(BaseModel):
+    entry: list[ScopusEntry] = []
+    total_results: str | None = Field(alias="opensearch:totalResults", default=None)
+    items_per_page: str | None = Field(alias="opensearch:itemsPerPage", default=None)
+
+
+class ScopusSearchResponse(BaseModel):
+    search_results: Optional[ScopusSearch] = Field(alias="search-results", default=None)
```
Comment (CreativePhilip) [inline rms/search_engine/models/scopus_models.py:56]:
```suggestion
    search_results: ScopusSearch | None = Field(alias="search-results", default=None)
```

**Review by CreativePhilip — CHANGES_REQUESTED** [review]

File: rms/search_engine/clients.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            title_criteria = " OR ".join([f'TITLE("{part}")' for part in title_parts])
+            search_criteria.append(f"({title_criteria})")
+
+        if query.keywords:
+            keywords_criteria = " OR ".join([f'KEY("{keyword}")' for keyword in query.keywords])
+            search_criteria.append(f"({keywords_criteria})")
+
+        if query.abstract_keywords:
+            abstract_keywords_criteria = " OR ".join([f'ABS("{keyword}")' for keyword in query.abstract_keywords])
+            search_criteria.append(f"({abstract_keywords_criteria})")
```
Comment (CreativePhilip) [inline rms/search_engine/clients.py:37]: I would separate this into 3 private method, each of which would handle the check(`if query.title` for example) inside. Afterwards the main function body can get the search criteria in just 3 linear fn calls

**Review by CreativePhilip — CHANGES_REQUESTED** [review]

File: rms/search_engine/clients.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    def __init__(self):
+        self.settings = Settings()
+        self.headers = {
+            "httpAccept": "application/json",
+            "X-ELS-APIKey": self.settings.scopus_api_key
+        }
+        self.scopus_search_api_endpoint = self.settings.scopus_search_endpoint
+
+
+class ScopusArticleSearchApi(ScopusClient):
```
Comment (CreativePhilip) [inline rms/search_engine/clients.py:20]: One stylistic comment regarding the class.
Please order the methods in such a way (`__init__`, public methods, private methods, other magic methods)

**Review by CreativePhilip — CHANGES_REQUESTED** [review]

**Review by CreativePhilip — APPROVED** [review]

File: rms/search_engine/utils.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import os
+
+
+class StopWordsProcessor:
+    STOPWORDS_PATH = os.path.join(os.path.dirname(__file__), 'assets', 'stopwords.txt')
```
Comment (CreativePhilip) [inline rms/search_engine/utils.py:5]: Please use pathlib for these operations

File: rms/search_engine/utils.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import os
+
+
+class StopWordsProcessor:
```
Comment (CreativePhilip) [inline rms/search_engine/utils.py:4]: For now this is fine, but for the future we would like to cache the wordlist between instance creations

**Review by CreativePhilip — CHANGES_REQUESTED** [review]

**Review by CreativePhilip — APPROVED** [review]

**Pull request merged**